Elly Vending step6

dely2p opened the pull request:
swift Vending step6 입니다.

dely2p committed:
fix UI to add position purchase list #35

구매 목록 이미지를 화면 아래 이미지로 추가하기 위해 UI를 수정했다.

dely2p committed:
add Button to purchase for user #37

구매 버튼을 collection button으로 추가함

dely2p committed:
add buyBeverage() method in ViewController and fix parameter to get beverageName #37

ViewController에 buyBeverage() method를 추가했다. 그리고 매개변수타입을 변경했다.

dely2p committed:
create PurchaseProduct class to data object transform #37

구매일자와 구매음료를 포함한 PurchaseProduct class를 만들어서 구매내역에서 사용하려함

dely2p committed:
add Notification to purchase beverage signal #37

음료 구매 시 알림 추가하는 기능을 만들었다.

dely2p committed:
add UIImageView when get notification #38

notification을 받으면 UIImageView를 수동으로 추가하도록 만들었다.

dely2p committed:
add imageName each Beverage Class #38

이미지 이름을 Coffee클래스와 하위 클래스에 초기화 함

dely2p committed:
add ImageMapper protocol with extension to get image name #38

클래스 이름을 기반으로 앞글자만 소문자로 바꾸고 뒤에 .jpg를 붙여서 이미지를 불러온다.

dely2p committed:
archive and add imageView for initialize scene #35

실행 이후 구매 목록을 화면 아래 이미지로 추가하도록 만들었다.

dely2p committed:
remove imageFromMenu() in InventoryMenu #36

필요없는 InventoryMenu imageFromMenu() 함수를 지움

dely2p committed:
add addBalance button to coin #39

addBalance Button 추가

dely2p committed:
add Coin Notification #39

코인 추가 알림을 넣었다.

dely2p committed:
fix readme.md file and add price in initialize of each Beverage #39

readme.md file step6 실행화면 추가 및 각 음료 객체들의 가격 추가함

godrm commented on VendingMachineApp/VendingMachineApp/BeverageGroup/MilkGroup/ChocoMilk.swift:
```diff
@@ -11,7 +11,7 @@ import Foundation
     private var cocoaPowder: Double = 1.1
     convenience init() {
-        self.init(brand: "", weight: 0, price: 0, name: "", manufactureDate: Date(), cocoaPower: 1)
+        self.init(brand: "", weight: 0, price: 1200, name: "", manufactureDate: Date(), cocoaPower: 1)
```
하드코딩되어 있는 숫자가 눈에 띄네요. 코코아 파우더 1은 뭐고 1.1은 어떤 의미죠?

dely2p replied:
하드코딩했던 것을 BeverageData라는 struct를 만들어서 data를 넣어두었습니다. (코코아파우더 1은 잘못된 하드코딩 초기화 입니다;;)

godrm commented on VendingMachineApp/VendingMachineApp/Model/ImageMapper.swift:
```diff
@@ -0,0 +1,23 @@
+        let name = String(describing: type(of: self))
+        let lowercasedOffirstWord = String(name[name.startIndex]).lowercased()
+        let anotherWords = name.dropFirst()
+        let result = lowercasedOffirstWord + anotherWords + ".jpg"
```
우아.. 이미지 파일 매칭하려고 너무 많은 일을 하고 있는거 아닌가요?
그냥 클래스 파일명으로 이미지 파일 이름을 바꾸는 건 어때요? 이런 작업은 오버에요. 
관리도 쉽고, 코드로 나누고 편집하기 보다는 매칭을 쉽게 하는게 더 중요합니다.

dely2p replied:
안그래도 다 만들고는 이미지 파일 이름을 바꿀걸.. 하고는 생각했지만 만들었던 게 아까워서 썼습니다..ㅠㅠ 
이미지 파일 이름 수정하여 매칭했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -64,11 +64,13 @@ class VendingMachine: NSObject, NSCoding, Vending {
             self.coins -= beverageOfChoice.price
+            NotificationCenter.default.post(name: Notification.Name.DidResetBalance, object: self)
         } catch {
             print("Beverage error")
```
자판기 앱인데... 구매 실패하면 뭘 해야할까요? 단지 print()가 끝일까요?

dely2p replied:
print() 대신 alert를 사용하여 구매 실패를 알려주도록 개선했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -12,29 +12,45 @@ class ViewController: UIViewController {
     var vendingMachine: Vending!
     var inventoryBox = InventoryBox()
     typealias TypeOf = InventoryBox.InventoryMenu
+    var imageX = 0
```
ViewController 변수들은 왜 아직도 public인거죠?

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -67,11 +83,47 @@ class ViewController: UIViewController {
+        guard let vending = self.vendingMachine as? VendingMachine else {
+            return
+        }
+        VendingMachine.storedInstance(vending)
```
이거 뭐죠? 왜 vending 변수를 새로 바꾸는거죠?

dely2p replied:
이전에 Notification을 쓰지 않을 때 소스코드를 그대로 두고 사용했었어요;; 수정했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -12,29 +12,45 @@ class ViewController: UIViewController {
             imageOfMenu[index].layer.cornerRadius = imageOfMenu[index].frame.width/4
+            imageOfMenu[index].layer.borderWidth = 5
             imageOfMenu[index].clipsToBounds = true
         }
```
imageOfMenu 화면 설정하는 코드를 여기에 두지 않고 하위 모듈로 감추도록 개선해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -12,29 +12,45 @@ class ViewController: UIViewController {
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(changeCoin),
+                                               name: Notification.Name.DidResetBalance,
+                                               object: nil)
```
viewDidLoad 에서도 여러 작업을 하고 있는데, 한 가지 기능만 하는 여러 메소드로 분리하는 것도 방법입니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -67,11 +83,47 @@ class ViewController: UIViewController {
+        let imageView = ImageViewMaker.makeImageView(imageX: imageX)
+        imageView.image = image
+        self.view.addSubview(imageView)
+        self.imageX += 50
```
이 함수는 print-로 시작하고, 복수로 끝나는데 음료 하나만 처리하고 있어서 내부 구현과 차이가 있어 보입니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -12,29 +12,45 @@ class ViewController: UIViewController {
+            printPurchaseProducts(beverage: products.purchaseBeverage)
+        }
         let formatter = DateFormatter()
         formatter.dateFormat = "yyyyMMdd"
```
여기 Formatter는 뜬금없이 뭔가요?

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -67,11 +83,47 @@ class ViewController: UIViewController {
+    }
+    
+    @objc func changeCoin() {
+        balance.text = String(vendingMachine.checkBalance())
```
`원` 단위 같은건 안 붙이나요?

dely2p replied:
원 단위와 가격(숫자) 라벨을 Main storyboard에서 따로 만들었습니다.

godrm reviewed (commented):
피드백 내용을 확인해서 개선해보세요.

dely2p committed:
fix image name to match class name #41

이미지 이름 첫번째 문자를 대문자로 고쳐서 클래스 이름과 일치하도록 수정했다. 이를 ImageMapper에서 .jpg를 붙여서 사용하도록 개선함

dely2p committed:
add purchase failure alert #42

재고가 0인 음료를 구매하려 할 때 구매 실패 메시지 띄워지도록 만들었다.

dely2p committed:
fix viewController Access Control 43

viewController 접근제어 수정함

dely2p committed:
remove storedInstance() in ViewController #44

옵저버 적용하기 전에 사용했던 storedInstance()를 지움

dely2p committed:
add custom imageView to hide the viewController's code #45

뷰 컨트롤러의 imageView 라운딩과 보더 설정을 RoundImageView라는 커스텀 ImageView를 따로 만들어서 설정해두었다.

dely2p committed:
remove date format #48

이전 코드에서 사용하던 date format을 삭제함

dely2p committed:
fix method name from printPurchaseProducts() to addPurchaseImage #47

메소드 이름을 메소드 코드에 맞게 수정했다.

dely2p committed:
fix separate viewDidLoad() method #46

viewDidLoad() method 분리하기

dely2p committed:
add BeveragePrice struct to fix hard coding of each beverage class #40

각 음료의 가격을 작성해둔 BeveragePrice를 추가하고 각 beverage 클래스에서 사용함

godrm commented on VendingMachineApp/VendingMachineApp/Model/ImageMapper.swift:
```diff
@@ -15,9 +15,7 @@ protocol ImageMapper {
-        let lowercasedOffirstWord = String(name[name.startIndex]).lowercased()
-        let anotherWords = name.dropFirst()
-        let result = lowercasedOffirstWord + anotherWords + ".jpg"
+        let result = name + ".jpg"
```
이런건 테스트 코드로 작성해서 확인해보면 어떨까요? 
작은 기능이지만 테스트할 수 있다는 것은 중요합니다. (작성할 필요는없어요)

dely2p replied:
네:) 테스트 코드로 확인하는 게 좋은 것 같습니다. (사실.. .jpg를 붙이는 것을 깜빡하고 name만 넘겼더니 사진이 나오지 않아서 몇 분 동안 헤맸었어요ㅠㅠ)

godrm commented:
수고하셨습니다. 이번 단계는 마무리하세요 ~
